Use RemoteAST in GetDynamicTypeAndAddress()

slavapestov opened the pull request:
There was a lot of redundant logic here that predates RemoteAST.

I started off by trying to simplify debug info emission in the Swift side by only emitting debug values for top-level generic parameter types and not member types; in lldb we should be able to look up member types just by calling `Type::subst()`. Unfortunately, this broke test cases where we were getting the dynamic type of values whose type is a member type. Turns out this was handled as part of a special "generic parameter or member type" code path, which would map the type to a concrete type, and then get the dynamic type of the concrete type.

This "get dynamic type of concrete type resulting from concretizing a generic parameter type" code path was totally different from the "get dynamic type of concrete type" code path. The former used RemoteAST, the latter did not. Merging these two code paths is what this PR is all about.

The non-RemoteAST code path handled a few cases that the RemoteAST code path did not, such as values materialized in host memory -- this is how results come back from the expression evaluator, which wants to "freeze" them so that they're available later. It also handled imported types and some other edge cases.

Now that RemoteAST on the Swift side is more capable, all it took was adding some support to lldb's MemoryReader implementation for reading memory from a specified local buffer to use RemoteAST for all the other cases as well.

Fixes <rdar://40706055> (but really a whole lot more than that).

slavapestov committed:
[Swift] Use Type::subst() instead of Type::transform() to substitute generic parameters

This will allow a corresponding simplification on the IRGen side,
since we no longer need to emit debug info for dependent member
types.

slavapestov committed:
[Swift] Merge GetDynamicTypeAndAddress_Struct() and GetDynamicTypeAndAddress_Enum()

These two functions are identical, and don't do anything specific
to structs or enums, so let's combine them into one more general
function.

slavapestov committed:
[Swift] Remove GetDynamicTypeAndAddress_Tuple()

It's not really correct to unwrap existentials in tuple elements because
the resulting concrete tuple type has a different layout than the original
type.

This changes some test output but note that the old behavior was
inconsistent; we would only call GetDynamicTypeAndAddress_Tuple() if the
original type was already generic, so a type like (Error, T) for a generic
parameter 'T' would unwrap the Error payload, whereas a type like
(Error, Int) would not.

slavapestov committed:
[Swift] Remove earlier workaround for tuple dynamic type issue

slavapestov committed:
[Swift] Add escape hatch allowing RemoteAST to read locally materialized values

slavapestov committed:
[Swift] Refactor GetDynamicTypeAndAddress_Protocol() to use RemoteAST

slavapestov committed:
[Swift] Clean up SwiftASTContext::IsPossibleDynamicType()

slavapestov committed:
[Swift] Better tagged pointer test

slavapestov committed:
[Swift] Remove tagged pointer check from SwiftLanguageRuntime::GetDynamicTypeAndAddress_Class()

slavapestov committed:
[Swift] GetDynamicTypeAndAddress() checks for class or protocol after substitution

This will allow us to completely eliminate the generic parameter
path.

slavapestov committed:
[Swift] Remove GetDynamicTypeAndAddress_GenericTypeParam()

dcci commented on source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -1729,380 +1734,88 @@ bool SwiftLanguageRuntime::IsValidErrorValue(
+    // our MemoryReader know where it is so that RemoteAST can read
+    // from it. Also, there's something going on with Error values
+    // I don't quite understand.
+    is_error = GetSwiftType(protocol_type)
```
This is the precise reason I didn't switch to RemoteAST this codepath after adding the support in the compiler. Apparently the expression parser does something weird with materializing/dematerializing values, @jimingham has a rdar assigned to him about this.

slavapestov replied:
For anyone else internal the radar is rdar://38194957. While it would be nice to fix this it's not a pre-requisite since we can teach the RemoteAST memory reader to read values from known local ranges of memory too. It would clean up the code here if we didn't have to worry about that, though.

slavapestov committed:
[Swift] Remove unnecessary GetDynamicValue() calls from materializer

They were actually unnecessary and doing the wrong thing when we had generic
parameters wrapping existentials.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov committed:
[Swift] Remove dead code from SwiftLanguageRuntime::CalculateErrorValue()

slavapestov committed:
[Swift] Add level of indirection in SwiftLanguageRuntime::CalculateErrorValue()

Existentials are represented as addresses of values, so a Swift error
existential should be the address of a pointer to an Error container
and not a pointer to the Error container itself.

slavapestov commented:
@dcci I figured out the problem with Error existentials. It was in SwiftLanguageRuntime::CalculateErrorValue(). There was also a RemoteAST bug on Linux with error values. This should now be ready to merge.

slavapestov commented:
https://github.com/apple/swift/pull/22975
@swift-ci Please test

dcci reviewed (approved):
LGTM. Thanks.

slavapestov commented:
@swift-ci Please test
